File: next/kmp/window/src/commonMain/kotlin/org/dweb_browser/sys/window/core/modal/WindowModalController.kt
package org.dweb_browser.sys.window.core.modal

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.flow.SharedFlow
import org.dweb_browser.core.module.NativeMicroModule
import org.dweb_browser.core.std.dns.nativeFetch
import org.dweb_browser.helper.Signal
import org.dweb_browser.helper.SimpleSignal
import org.dweb_browser.helper.buildUrlString

sealed class WindowModalController(
  val mm: NativeMicroModule.NativeRuntime,
  val modal: ModalState,
  val wid: String,
  onCallback: SharedFlow<ModalCallback>,
) {
  protected val onOpenSignal = SimpleSignal()
  val onOpen = onOpenSignal.toListener()
  protected val onCloseSignal = SimpleSignal()
  val onClose = onCloseSignal.toListener()
  protected val onDestroySignal = SimpleSignal()
  val onDestroy = onDestroySignal.toListener()
  protected val onStateChangeSignal = Signal<WindowModalState>()
  val onStateChange = onStateChangeSignal.toListener()
  protected var _state = WindowModalState.INIT
  protected fun isState(vararg states: WindowModalState) = states.contains(_state)
  protected inline suspend fun awaitState(crossinline isComplete: (WindowModalState) -> Boolean) =
    CompletableDeferred<Unit>().also { waiter ->
      val off = onStateChange {
        if (isComplete(it)) {
          waiter.complete(Unit)
          offListener()
        }
      }
      if (isComplete(_state)) {
        waiter.complete(Unit)
        off()
      }
    }.await()

  var state
    get() = _state
    protected set(value) {
      if (_state == value) {
        return
      }
      _state = value
      // 监听状态更新并且触发相关事件
      mm.scopeLaunch(cancelable = false) {
        onStateChangeSignal.emit(value)
        when (_state) {
          WindowModalState.OPEN -> onOpenSignal.emit(Unit)
          WindowModalState.CLOSE -> onCloseSignal.emit(Unit)
          WindowModalState.DESTROY -> {
            onCloseSignal.emitAndClear()
            onDestroySignal.emitAndClear()
            onOpenSignal.clear()
          }

          else -> {}
        }
      }
    }

  init {
    // 在回调事件触发的时候更新state
    mm.scopeLaunch(cancelable = true) {
      onCallback.collect {
        state = when (it) {
          is CloseAlertModalCallback -> WindowModalState.CLOSE
          is CloseModalCallback -> WindowModalState.CLOSE
          is OpenModalCallback -> WindowModalState.OPEN
          is DestroyModalCallback -> WindowModalState.DESTROY
        }
      }
    }
  }

  val isDestroyed get() = isState(WindowModalState.DESTROYING, WindowModalState.DESTROY)
  suspend fun open() {
    if (isDestroyed || isState(WindowModalState.OPEN, WindowModalState.OPENING)) {
      return
    }
    state = WindowModalState.OPENING
    mm.nativeFetch(buildUrlString("file://window.sys.dweb/openModal") {
      parameters["modalId"] = modal.modalId
      parameters["wid"] = wid
    }).boolean()
  }

  suspend fun close() {
    if (isDestroyed || isState(WindowModalState.CLOSE, WindowModalState.CLOSING)) {
      return
    }
    state = WindowModalState.CLOSING
    mm.nativeFetch(buildUrlString("file://window.sys.dweb/closeModal") {
      parameters["modalId"] = modal.modalId
      parameters["wid"] = wid
    }).boolean()
  }

  suspend fun setCloseTip(closeTip: String?) {
    if (isDestroyed) {
      return
    }
    mm.nativeFetch(buildUrlString("file://window.sys.dweb/updateModalCloseTip") {
      parameters["modalId"] = modal.modalId
      closeTip?.let { parameters["closeTip"] = it }
    }).boolean()
  }

  suspend fun destroy() {
    if (isDestroyed) {
      return
    }
    this.state = WindowModalState.DESTROYING
    mm.nativeFetch(buildUrlString("file://window.sys.dweb/removeModal") {
      parameters["modalId"] = modal.modalId
    }).boolean()
    this.state = WindowModalState.DESTROY;
  }
}

File: toolkit/plaoc/plugins/components/window/WindowModalController.ts
import { PromiseOut } from "@dweb-browser/helper/PromiseOut.ts";
import { SafeEvent, SafeEventTarget, SafeStateEvent } from "../../helper/SafeEventTarget.ts";
import type { $Callback, Signal } from "../../helper/createSignal.ts";
import { BasePlugin } from "../base/base.plugin.ts";
import type { $Modal, $ModalCallback } from "./window.type.ts";

export const enum WindowModalState {
  INIT = "init",
  OPENING = "opening",
  OPEN = "open",
  CLOSING = "closing",
  CLOSE = "close",
  DESTROYING = "destroying",
  DESTROY = "destroy",
}
type $ModalEvents = {
  close: SafeEvent<WindowModalState.CLOSE>;
  open: SafeEvent<WindowModalState.OPEN>;
  destroy: SafeEvent<WindowModalState.DESTROY>;
  statechange: SafeStateEvent<"statechange", WindowModalState>;
};

export class WindowModalController extends SafeEventTarget<$ModalEvents> {
  constructor(readonly plugin: BasePlugin, readonly modal: $Modal, onCallback: Signal<$Callback<[$ModalCallback]>>) {
    super();
    onCallback.listen((callbackData) => {
      switch (callbackData.type) {
        case "open":
          this.setState(WindowModalState.OPEN);
          break;
        case "close":
          this.setState(WindowModalState.CLOSE);
          break;
        case "close-alert":
          this.setState(WindowModalState.CLOSE);
          break;
      }
    });
    this._init();
  }
  protected _init() {}
  #state = WindowModalState.INIT;
  private setState(value: WindowModalState) {
    if (value === this.#state) {
      return;
    }
    this.#state = value;
    this.dispatchEvent(new SafeStateEvent("statechange", { state: value }));
    switch (value) {
      case WindowModalState.OPEN:
        this.dispatchEvent(new SafeEvent(WindowModalState.OPEN));
        break;
      case WindowModalState.CLOSE:
        this.dispatchEvent(new SafeEvent(WindowModalState.CLOSE));
        break;
      case WindowModalState.DESTROY:
        this.dispatchEvent(new SafeEvent(WindowModalState.DESTROY));
        break;
    }
  }
  get state() {
    return this.#state;
  }
  private isState(...states: WindowModalState[]) {
    for (const state of states) {
      if (state === this.state) {
        return true;
      }
    }
    return false;
  }
  protected awaitState(waitState: WindowModalState) {
    if (this.state === waitState) {
      return;
    }
    const waiter = new PromiseOut<void>();
    // deno-lint-ignore no-this-alias
    const self = this;
    this.addEventListener("statechange", function statechange(event) {
      if (event.state === waitState) {
        waiter.resolve();
        self.removeEventListener("statechange", statechange);
      }
    });
    return waiter.promise;
  }
  get isDestroyed() {
    return this.isState(WindowModalState.DESTROYING, WindowModalState.DESTROY);
  }
  async open() {
    if (this.isDestroyed || this.isState(WindowModalState.OPENING, WindowModalState.OPEN)) {
      return;
    }
    this.setState(WindowModalState.OPENING);
    await this.plugin
      .fetchApi("/openModal", { pathPrefix: "window.sys.dweb", search: { modalId: this.modal.modalId } })
      .boolean();
    await this.awaitState(WindowModalState.OPEN);
  }
  async close() {
    if (this.isDestroyed || this.isState(WindowModalState.CLOSING, WindowModalState.CLOSE)) {
      return;
    }
    this.setState(WindowModalState.CLOSING);
    await this.plugin
      .fetchApi("/closeModal", { pathPrefix: "window.sys.dweb", search: { modalId: this.modal.modalId } })
      .boolean();
    await this.awaitState(WindowModalState.CLOSE);
  }
  async destroy() {
    if (this.isDestroyed) {
      return;
    }
    this.setState(WindowModalState.DESTROYING);
    await this.plugin
      .fetchApi("/removeModal", { pathPrefix: "window.sys.dweb", search: { modalId: this.modal.modalId } })
      .boolean();
    await this.awaitState(WindowModalState.DESTROY);
  }
}
